### DISCUSSION TITLE
introduction-to-authentication-with-asp-net-core/

### DISCUSSION THREAD

**Discussion by andrewlock** [andrewlock/blog-comments#153] [Announcements]
# introduction-to-authentication-with-asp-net-core/

This is the first in a series of posts on authentication and authorisation. In this post, I talk about claims-based authentication works in ASP.NET Core.

https://andrewlock.net/introduction-to-authentication-with-asp-net-core/

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at August 12 2016, 08:45, in reply to [Legolas](https://disqus.com/by/legolss/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/153#discussioncomment-5795267)</em>

---

<p>Legolas thanks, appreciate it.</p><p>The urls used with claim types do not actually have to exist, and as you've seen in this case, it doesn't! They simply need to be unique for a particular claim value, so urls make a good fit.</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at August 14 2016, 10:19, in reply to [Andriy Tolstoy](https://disqus.com/by/andriytolstoy/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/153#discussioncomment-5795272)</em>

---

<p>Thanks, fixed.</p><p>ClaimTypes.Name is the username of the user, so for  example in twitter it might be 'andrewlocknet'<br>ClaimTypes.NameIdentifier is the identifier of the user, e.g. it might be a userId of 42342, or a GUID</p><p>Hope that helps!</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at August 14 2016, 11:21, in reply to [Mark Redman](https://disqus.com/by/disqus_wDI1DufMbR/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/153#discussioncomment-5795274)</em>

---

<p>Hi Mark,</p><p>I think you've hit on some of the biggest difficulties of getting this approach to work, I suspect most people working on non-trivial applications have found the same thing.</p><p>First off, I don't think roles in of themselves are 'bad', it's more about how they're used. When I talked about using 'IsInRole' as being for backward compatibility reasons, it's primarily this use of roles which I think is being frowned upon.</p><p>Instead, you probably want to try and separate the roles a user is in, from what those roles are allowed to do. Ben Collins has an interesting post in which he talks about using a PermissionsEnum to authorise users. Those permissions could be granted based on the roles a user is in, and those could be stored on the user as claims, depending on your requirements. The link is here: <a href="http://benjamincollins.com/blog/practical-permission-based-authorization-in-asp-net-core/" rel="nofollow noopener" title="http://benjamincollins.com/blog/practical-permission-based-authorization-in-asp-net-core/">http://benjamincollins.com/blog/practical-permission-based-authorization-in-asp-net-core/</a></p><p>One thing I would advocate is the use of requirements and policies. I'm going to do a post on these soon, but basically they allow you to define orthogonal requirements, independent of specific roles and claims etc.</p><p>For example, you talk about an 'IsAdministrator' authorisation, but I would avoid that sort of thing. Rather, you are authorising that a user is allowed to, for example, create a new user. Instead, I would define a policy 'CanCreateUser'. Under the hood, that might simply check for the Administrator role, or it could check that the user has also confirmed their email address, of any other requirement.</p><p>The point is that roles are not directly used for authorisation. A role is just another claim. Requirements and policies are used instead.</p><p>Hope that helps, I'm still not entirely sure the best approach, but I think that gives a good combination of results. If that's not clear I'll try and work it up into a post instead:)</p>

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at January 12 2017, 10:06, in reply to [Ahmad Moussawi](https://disqus.com/by/ahmadmoussawi/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/153#discussioncomment-5795277)</em>

---

<p>Yep, I completely agree - prefer claims wherever possible!</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at September 17 2016, 08:52, in reply to [Felix Rabinovich](https://disqus.com/by/felixrabinovich/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/153#discussioncomment-5795283)</em>

---

<p>Thanks, fixed </p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at September 29 2016, 10:26, in reply to [Strimilly](https://disqus.com/by/strimilly/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/153#discussioncomment-5795287)</em>

---

<p>Hi, thanks:)</p><p>You are correct about ExpiresUTC - it specifies the date the cookie expires. If the cookie expires, it will be removed and you will have to login again.</p><p>The way around this, as you suspected, is using AllowRefresh. If you are more than halfway through the 'life' of the ticket, AllowRefresh = true and SlidingExpiration = true, then the Issued and Expires time of the authentication ticket will be updated, and your session will continue.</p><p>Hope that makes sense!:)</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at October 28 2016, 01:41, in reply to [Dmitry Novoselov](https://disqus.com/by/dmitry_novoselov/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/153#discussioncomment-5795293)</em>

---

<p>Glad you figured it out - yes the order of middleware is critical!</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at November 10 2016, 09:52, in reply to [Sam Piper](https://disqus.com/by/disqus_tRkX8luHUZ/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/153#discussioncomment-5795298)</em>

---

<p>Thanks for reading. Yes, if you set ShouldRenew = true, a refresh cookie will be set, assuming it passes your validation handler. You might find my post on Cookie Authentication useful: <a href="http://andrewlock.net/exploring-the-cookieauthenticationmiddleware-in-asp-net-core/#thecookieauthenticationhandler" rel="nofollow noopener" title="http://andrewlock.net/exploring-the-cookieauthenticationmiddleware-in-asp-net-core/#thecookieauthenticationhandler">http://andrewlock.net/exploring-the-cookieauthenticationmiddleware-in-asp-net-core/#thecookieauthenticationhandler</a></p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at May 12 2017, 08:49, in reply to [ferr ferr](https://disqus.com/by/ferr_ferr/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/153#discussioncomment-5795572)</em>

---

<p>Nothing is built into ASP.NET Core itself, but it is built into ASP.NET Core Identity. This creates AspNetRoleClaims and AspNetUserClaims tables so you can add (and manage) the necessary claims for a particular user.</p><p>The management and persistence of claims is completely separate from the details in this post though. I recommend checking out IdentityServer 4 if that's something you're looking into. <a href="http://docs.identityserver.io/" rel="nofollow noopener" title="http://docs.identityserver.io/">http://docs.identityserver.io/</a></p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at July 14 2017, 11:44, in reply to [Steve Hibbert](https://disqus.com/by/steve_hibbert/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/153#discussioncomment-5795596)</em>

---

<p>Glad it makes sense :) But what do you mean "past love"? #noshame ;)</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at April 05 2018, 04:11, in reply to [Adeniranyusuf Abdullateef](https://disqus.com/by/adeniranyusufabdullateef/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/153#discussioncomment-5795607)</em>

---

<p>Cheers, I need to update the links... This is the last post in the series which includes links to the earlier ones: <a href="https://andrewlock.net/resource-specific-authorisation-in-asp-net-core/" rel="nofollow noopener" title="https://andrewlock.net/resource-specific-authorisation-in-asp-net-core/">https://andrewlock.net/resource-specific-authorisation-in-asp-net-core/</a></p>

[a comment by an automated account is not shown]

**Comment by PhilJollans** [comment]
Hi Andrew,  
since this is a series of posts, it would be nice if there was a list of links to all the posts in the series.

Phil
